src/algebra/parity.lean

```diff
@@ -342,3 +364,39 @@ end
 (even_bit0 _).abs_zpow _
 
 end field_power
+
+section decidability
+
+namespace nat
+
+instance : decidable_pred (is_square : ℕ → Prop)
+| n :=
+begin
+  convert decidable_of_iff' _ (nat.exists_mul_self n),
+  unfold is_square,
+  congr,
+  ext,
+  split,
+  { intros h, simp [h], },
+  { intros h, simp [h],},
+end
+
+end nat
+
+namespace int
+
+instance : decidable_pred (is_square : ℤ → Prop)
+| n :=
+begin
+  convert decidable_of_iff' _ (int.exists_mul_self n),
+  unfold is_square,
+  congr,
+  ext,
+  split,
+  { intros h, simp [h], },
+  { intros h, simp [h], },
+end
```

Comment:
Do NOT use `convert` when defining a decidability instance. 
```suggestion
λ n, decidable_of_iff' _ $ (exists_congr $ λ _, eq_comm).trans int.exists_mul_self n
```
I am not sure whether using the equation compiler instead of a `λ` makes any difference here, but it's good practice to avoid it where you can.

Reply:
Oh sure! I should have guessed it was `exists_congr` but couldn't think of it.

Also, why shouldn't `convert` be used in decidability instances? Just curious!

Reply:
Hmm, this proof isn't working for me in `data.nat.parity`. Specifically, I have to hint it as follows:

```lean
instance : decidable_pred (is_square : ℕ → Prop) :=
begin
  intros n,
  have : is_square n ↔ ∃ (n' : ℕ), n' * n' = n,
  { exact exists_congr (λ _, eq_comm), },
  exact decidable_of_iff' _ (this.trans n.exists_mul_self),
end
```

If I just do:

```lean
instance : decidable_pred (is_square : ℕ → Prop) :=
λ n, decidable_of_iff' _ $ (exists_congr $ λ _, eq_comm).trans n.exists_mul_self
```

I get the error:

```
type mismatch at application
  (exists_congr (λ (_x : ?m_1), eq_comm)).trans (exists_mul_self n)
term
  exists_mul_self n
has type
  (∃ (n_1 : ℕ), n_1 * n_1 = n) ↔ sqrt n * sqrt n = n
but is expected to have type
  (∃ (a : ?m_1), ?m_3 = ?m_4) ↔ ?m_5
```

Reply:
`convert` might mess up with the definitional unfolding of the resulting term. As `decidable` is data, this matters to avoid weird errors where things are "visibly the same" but hidden instances do not match.
About my proof, yeah annoyingly elaboration breaks it. I believe this should work instead:
```
instance : decidable_pred (is_square : ℕ → Prop) :=
λ n, decidable_of_iff _ $ n.exists_mul_self.symm.trans $ exists_congr $ λ _, eq_comm
```

Reply:
Gotcha! And these proofs have been updated.